### ISSUE TITLE
typedef "call-signature" and "parameter" trigger on array of function

### ISSUE THREAD

**Issue by nixxquality** [palantir/tslint#3524]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.5.3
- __Running TSLint via__: gulp-tslint (has type information)

#### TypeScript code being linted

```ts
type Operation = (a: number, b: number) => number;
const Operations: Operation[] = [];

Operations.push(function add(a, b) {
    return a + b;
});

Operations.push(function subtract(a, b) {
    return a - b;
});
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "typedef": true
    }
}
```

#### Actual behavior
```
ERROR: (typedef) Example.ts[5, 26]: expected call-signature: 'add' to have a typedef
ERROR: (typedef) Example.ts[5, 30]: expected parameter: 'a' to have a typedef
ERROR: (typedef) Example.ts[5, 33]: expected parameter: 'b' to have a typedef
ERROR: (typedef) Example.ts[9, 26]: expected call-signature: 'subtract' to have a typedef
ERROR: (typedef) Example.ts[9, 35]: expected parameter: 'a' to have a typedef
ERROR: (typedef) Example.ts[9, 38]: expected parameter: 'b' to have a typedef
```

#### Expected behavior
Type definition should be inferred from the type of the array.
TypeScript itself does this properly and displays correct type hints.

**Comment by ajafff** [comment]
Please note that this rule is not about what could be inferred. It's used to enforce type annotations even if they are not strictly necessary.
If you only want to enforce type annotations where necessary, you should not use this rule and enable `--noImplicitAny` in your compilerOptions instead.
